Concurrent Crawl Limit

ikreymer opened the pull request:
This PR covers both backend and frontend side of crawl concurrency per Organization.
It introduces a new crawl status, `waiting_org_list` and renames previous `waiting` -> `waiting_capacity`
The frontend displays this state as follows:
<img width="1180" alt="Screen Shot 2023-05-25 at 11 11 43 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/6b5f8ecb-78fc-4e76-98e4-f1657d190916">

The limit is set on the Organization object, as `maxParallelCrawls` field on the new `quotas` field, now included in each Organization.
The `org/quotas` endpoint allows configuring the quotas object.

The superadmin org page now has quotas popup, where all the available quotas can be configured:
<img width="692" alt="Screen Shot 2023-05-25 at 11 13 13 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/a1964ec1-6e24-4b9d-8d00-3c8b773da1ae">

Clicking the gear will show the popup, listing all the available quotas:
<img width="901" alt="Screen Shot 2023-05-25 at 11 14 06 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/8bf81e93-4571-4f47-bdd2-92e9a9f72984">

The `maxParallelCrawls` is the only field so far, adding additional fields to `quotas` on the backend will automatically expose them in this popup.
 
### Backend Implementation

The parallel crawl limit is enforced by the operator per organization. As part of this, the orgid is included as part of each crawljob. The operator sync method now includes all the running crawljobs for the org (sorted by creationTimestamp by default), and if quota is set, checks to see if each job is <=N order, and any jobs >N are set to `waiting_org_limit`.
(Canceled / not running jobs are excluded from the count).

ikreymer committed:
concurrent crawl limits: (addresses #866)
- support limits on concurrent crawls that can be run within a single org
- change 'waiting' state to 'waiting_org_limit' for concurrent crawl limit and 'waiting_capacity' for capacity-based
limits
- frontend: show different waiting states on frontend: 'Waiting (Crawl Limit) and 'Waiting (At Capacity)'
- operator: add all crawljobs as related, appear to be returned in creation order
- operator: if concurrent crawl limit set, ensures current job is in the first N set of crawljobs (as provided via 'related' list of crawljob objects)
 before it can proceed to 'starting', otherwise set to 'waiting_org_limit'
- api: add org /quotas endpoint for configuring quotas
- remove 'new' state, always start with 'starting'
- crawljob: add 'oid' to crawljob spec and label for easier querying

ikreymer committed:
fix merge issue

ikreymer committed:
tests: add concurrent crawl limit tests, setting limit to 1

ikreymer committed:
tests: fix state waiting -> waiting_capacity

tw4l commented on backend/btrixcloud/crawls.py:
```diff
@@ -599,7 +604,7 @@ async def update_crawl_state(self, crawl_id: str, state: str):
             {
                 "_id": crawl_id,
-                "state": {"$in": ["running", "starting", "canceling", "stopping"]},
+                "state": {"$in": RUNNING_AND_STARTING_STATES},
```
Do we want to remove "canceling" and "stopping"? Are those still states we have available? I don't see them in the `*_STATES` constants above.

ikreymer replied:
Yes, `stopping` is no longer a state, but a separate flag on the crawl, which indicates manually requested stop.
The crawl can still go through several phases (to be added) once its marked as stopping.
Didn't add a canceling state for now, as it should cancel fairly quickly / transition to 'canceled' state right away.
If needed, could handle it similar to stopping, but I think its fairly quick usual.

tw4l commented on backend/btrixcloud/orgs.py:
```diff
@@ -447,9 +471,10 @@ async def create_org(
+            return {"added": False, "error": "already_exists"}
 
-        return {"added": True}
+        return {"id": id_, "added": True}
```
Elsewhere we do `{"added": {"id": id_}}` or `{"added": id_}` - maybe we should try to be more consistent...

ikreymer replied:
yeah, i think we should do a general API consistency pass on this and other conventions

ikreymer replied:
Created #900 to track this.

ikreymer committed:
move concurrent crawl limit into Quotas object
tests: add new nightly test for limit

ikreymer committed:
admin org page: add initial popup for setting org quotas

ikreymer committed:
Merge branch 'main' into concurrent-crawl-limit

ikreymer committed:
readd NON_RUNNING_STATES

ikreymer committed:
don't localize quota key

ikreymer committed:
tests: lower sleep time, cancel 2nd crawl at end of tests

ikreymer committed:
Merge branch 'main' into concurrent-crawl-limit

ikreymer committed:
fix issues, don't set state before set_state

ikreymer committed:
fix canceling missing crawls

ikreymer committed:
ci: add logging of operator

ikreymer committed:
operator: more stringent state transitions: add allowed_from to set_state()
to ensure state transitions only happened from allowed states

ikreymer committed:
operator: fix allowed_states, just use lists, get current state if not changed

ikreymer committed:
ensure finished always set in mark_finished

ikreymer committed:
operator:
- ensure finished also synched along with state from db
- add indices by oid and cid
- log only invalid state transitions, ignore same state

ikreymer committed:
format fix

ikreymer committed:
ensure canceled/failed can be set from any state

tw4l reviewed (approved):
Small comments

ikreymer committed:
rename maxParallelCrawls -> maxConcurrentCrawls
add state transitions to set_state
